File: backend/app/routers/rag_stream.py
"""SSE streaming endpoint for real-time agent state updates."""
import asyncio
import uuid
import logging
from typing import Optional, Dict, Any

from fastapi import APIRouter, Request
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from app.rag_agent.callbacks import (
    StreamingCallbackHandler,
    get_callback_handler,
    remove_callback_handler,
    StepName
)
from app.rag_agent.graph import graph
from app.rag_agent.schemas.outputs import ChunkAnalysis


logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/rag", tags=["rag-stream"])


class StreamQueryRequest(BaseModel):
    """Request body for streaming RAG query."""
    query: str = Field(..., description="The query to process through the RAG pipeline")
    filters: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Optional metadata filters for retrieval"
    )


async def event_generator(
    session_id: str,
    query: str,
    filters: Optional[Dict[str, Any]],
    request: Request
):
    """
    Generate SSE events for the RAG pipeline execution.
    
    This function runs the RAG graph and emits progress events through
    the callback handler's event queue.
    """
    handler = get_callback_handler(session_id)
    
    try:
        # Emit pipeline started
        await handler.on_pipeline_start(query)
        yield handler.event_queue.get_nowait().to_sse() if not handler.event_queue.empty() else ""
        
        # Emit retrieve step started
        await handler.on_step_start(StepName.RETRIEVE)
        
        # Drain the queue after each emit
        while not handler.event_queue.empty():
            event = handler.event_queue.get_nowait()
            if event:
                yield event.to_sse()
        
        # Build initial state
        initial_state = {
            "query": query,
            "filters": filters,
            "retrieved_chunks": [],
            "analyses": [],
            "shared_document": "",
            "final_response": "",
            "retrieval_count": 0,
            "relevant_count": 0,
            "sources_used": []
        }
        
        # Stream through the graph execution
        # We use astream_events to get granular progress
        current_step = StepName.RETRIEVE
        analysts_tracked: set = set()
        
        async for event in graph.astream_events(initial_state, version="v2"):
            # Check if client disconnected
            if await request.is_disconnected():
                logger.info(f"Client disconnected for session {session_id}")
                break
            
            event_type = event.get("event")
            event_name = event.get("name", "")
            
            # Track node starts
            if event_type == "on_chain_start":
                if event_name == "retrieve":
                    # Already emitted above
                    pass
                elif event_name == "analyze_chunk":
                    # Track analyst start - happens in parallel
                    run_id = event.get("run_id", "")
                    if run_id and run_id not in analysts_tracked:
                        analysts_tracked.add(run_id)
                elif event_name == "synthesize":
                    await handler.on_step_start(StepName.SYNTHESIZE)
            
            # Track node completions
            elif event_type == "on_chain_end":
                output = event.get("data", {}).get("output", {})
                
                if event_name == "retrieve":
                    # Retrieve completed
                    retrieved_chunks = output.get("retrieved_chunks", [])
                    retrieval_count = len(retrieved_chunks)
                    
                    await handler.on_step_complete(
                        StepName.RETRIEVE,
                        {"retrieval_count": retrieval_count}
                    )
                    
                    # Emit distribute step
                    await handler.on_step_start(StepName.DISTRIBUTE)
                    
                    # Get chunk IDs for spawning
                    chunk_ids = [f"chunk_{i}" for i in range(retrieval_count)]
                    await handler.on_analysts_spawned(retrieval_count, chunk_ids)
                    
                    await handler.on_step_complete(StepName.DISTRIBUTE)
                    
                    # Start analyze phase
                    if retrieval_count > 0:
                        await handler.on_step_start(
                            StepName.ANALYZE,
                            {"total_analysts": retrieval_count}
                        )
                    current_step = StepName.ANALYZE
                
                elif event_name == "analyze_chunk":
                    # Individual analyst completed
                    analyses = output.get("analyses", [])
                    if analyses:
                        for analysis in analyses:
                            # Handle both dict and ChunkAnalysis objects
                            if isinstance(analysis, dict):
                                await handler.on_analyst_complete(
                                    analyst_id=analysis.get("chunk_id", "unknown"),
                                    chunk_id=analysis.get("chunk_id", "unknown"),
                                    relevance_score=analysis.get("relevance_score", 0),
                                    is_relevant=analysis.get("is_relevant", False),
                                    summary=analysis.get("summary", ""),
                                    confidence=analysis.get("confidence", "low")
                                )
                            else:
                                await handler.on_analyst_complete(
                                    analyst_id=analysis.chunk_id,
                                    chunk_id=analysis.chunk_id,
                                    relevance_score=analysis.relevance_score,
                                    is_relevant=analysis.is_relevant,
                                    summary=analysis.summary,
                                    confidence=analysis.confidence
                                )
                
                elif event_name == "synthesize":
                    # Synthesis completed - emit final response
                    await handler.on_step_complete(StepName.SYNTHESIZE)
                    
                    await handler.on_final_response(
                        final_response=output.get("final_response", ""),
                        retrieval_count=output.get("retrieval_count", 0),
                        relevant_count=output.get("relevant_count", 0),
                        sources_used=output.get("sources_used", []),
                        shared_document=output.get("shared_document", "")
                    )
            
            # Drain event queue after each graph event
            while not handler.event_queue.empty():
                sse_event = handler.event_queue.get_nowait()
                if sse_event:
                    yield sse_event.to_sse()
        
        # Close handler when done
        await handler.close()
        
    except Exception as e:
        logger.error(f"Error in streaming RAG query: {e}")
        await handler.on_error(str(e), current_step.value if 'current_step' in dir() else None)
        
        # Emit error event
        while not handler.event_queue.empty():
            event = handler.event_queue.get_nowait()
            if event:
                yield event.to_sse()
        
        await handler.close()
    
    finally:
        # Cleanup
        remove_callback_handler(session_id)


@router.post("/stream")
async def stream_rag_query(request: Request, body: StreamQueryRequest):
    """
    Stream RAG pipeline execution with real-time progress events.
    
    This endpoint uses Server-Sent Events (SSE) to stream progress updates
    as the multi-agent RAG pipeline processes the query.
    
    Event Types:
    - pipeline_started: Query processing has begun
    - step_started: A pipeline step (retrieve/distribute/analyze/synthesize) has started
    - step_completed: A pipeline step has finished
    - analyst_spawned: Analyst agents have been created for parallel chunk analysis
    - analyst_completed: An individual analyst has finished analyzing a chunk
    - synthesis_started: Final synthesis step has begun
    - final_response: The final response is ready
    - error: An error occurred during processing
    
    Returns:
        StreamingResponse: SSE stream of AgentEvent objects
    """
    session_id = str(uuid.uuid4())
    
    return StreamingResponse(
        event_generator(session_id, body.query, body.filters, request),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",  # Disable nginx buffering
        }
    )


@router.get("/stream/health")
async def stream_health():
    """Health check for the streaming endpoint."""
    return {"status": "healthy", "streaming": True}

File: frontend/src/api/ragStream.ts
/**
 * SSE client utilities for consuming agent state stream.
 */

const API_BASE = '/api';

// Event types from the backend
export type EventType =
  | 'pipeline_started'
  | 'step_started'
  | 'step_completed'
  | 'analyst_spawned'
  | 'analyst_completed'
  | 'synthesis_started'
  | 'final_response'
  | 'error';

export type StepName = 'retrieve' | 'distribute' | 'analyze' | 'synthesize';
export type ConfidenceLevel = 'high' | 'medium' | 'low';

// Agent event from SSE stream
export interface AgentEvent {
  type: EventType;
  timestamp: string;
  data: {
    query?: string;
    step?: StepName;
    duration_ms?: number;
    retrieval_count?: number;
    total_analysts?: number;
    chunk_ids?: string[];
    analyst_id?: string;
    chunk_id?: string;
    relevance_score?: number;
    is_relevant?: boolean;
    summary?: string;
    confidence?: ConfidenceLevel;
    completed_analysts?: number;
    relevant_count?: number;
    final_response?: string;
    sources_used?: string[];
    shared_document?: string;
    total_duration_ms?: number;
    error?: string;
  };
}

// Pipeline state for UI
export type PipelineStatus =
  | 'idle'
  | 'retrieving'
  | 'distributing'
  | 'analyzing'
  | 'synthesizing'
  | 'complete'
  | 'error';

export type StepStatus = 'pending' | 'active' | 'complete';

export interface AnalystState {
  id: string;
  chunkId: string;
  status: 'pending' | 'analyzing' | 'complete';
  relevanceScore?: number;
  isRelevant?: boolean;
  summary?: string;
  confidence?: ConfidenceLevel;
}

export interface RetrievalState {
  status: StepStatus;
  chunksRetrieved: number;
  durationMs?: number;
}

export interface AnalysisState {
  status: StepStatus;
  totalAnalysts: number;
  completedAnalysts: number;
  analysts: AnalystState[];
  durationMs?: number;
}

export interface SynthesisState {
  status: StepStatus;
  durationMs?: number;
}

export interface PipelineResult {
  finalResponse: string;
  retrievalCount: number;
  relevantCount: number;
  sourcesUsed: string[];
  sharedDocument: string;
}

export interface PipelineState {
  status: PipelineStatus;
  currentStep: number; // 1-4
  totalSteps: 4;

  // Retrieval phase
  retrieval: RetrievalState;

  // Analysis phase
  analysis: AnalysisState;

  // Synthesis phase
  synthesis: SynthesisState;

  // Results
  result?: PipelineResult;

  // Error info
  error?: string;
  totalDurationMs?: number;
}

// Initial state factory
export function createInitialPipelineState(): PipelineState {
  return {
    status: 'idle',
    currentStep: 0,
    totalSteps: 4,
    retrieval: {
      status: 'pending',
      chunksRetrieved: 0,
    },
    analysis: {
      status: 'pending',
      totalAnalysts: 0,
      completedAnalysts: 0,
      analysts: [],
    },
    synthesis: {
      status: 'pending',
    },
  };
}

// State reducer for processing events
export function processAgentEvent(
  state: PipelineState,
  event: AgentEvent
): PipelineState {
  const newState = { ...state };

  switch (event.type) {
    case 'pipeline_started':
      return {
        ...createInitialPipelineState(),
        status: 'retrieving',
        currentStep: 1,
        retrieval: {
          status: 'active',
          chunksRetrieved: 0,
        },
      };

    case 'step_started':
      switch (event.data.step) {
        case 'retrieve':
          newState.status = 'retrieving';
          newState.currentStep = 1;
          newState.retrieval.status = 'active';
          break;
        case 'distribute':
          newState.status = 'distributing';
          newState.currentStep = 2;
          break;
        case 'analyze':
          newState.status = 'analyzing';
          newState.currentStep = 3;
          newState.analysis.status = 'active';
          if (event.data.total_analysts !== undefined) {
            newState.analysis.totalAnalysts = event.data.total_analysts;
          }
          break;
        case 'synthesize':
          newState.status = 'synthesizing';
          newState.currentStep = 4;
          newState.synthesis.status = 'active';
          break;
      }
      break;

    case 'step_completed':
      switch (event.data.step) {
        case 'retrieve':
          newState.retrieval.status = 'complete';
          newState.retrieval.durationMs = event.data.duration_ms;
          if (event.data.retrieval_count !== undefined) {
            newState.retrieval.chunksRetrieved = event.data.retrieval_count;
          }
          break;
        case 'distribute':
          // Distribution complete, moving to analyze
          break;
        case 'analyze':
          newState.analysis.status = 'complete';
          newState.analysis.durationMs = event.data.duration_ms;
          break;
        case 'synthesize':
          newState.synthesis.status = 'complete';
          newState.synthesis.durationMs = event.data.duration_ms;
          break;
      }
      break;

    case 'analyst_spawned':
      newState.analysis.totalAnalysts = event.data.total_analysts || 0;
      newState.analysis.analysts = (event.data.chunk_ids || []).map(
        (chunkId: string) => ({
          id: chunkId,
          chunkId,
          status: 'pending' as const,
        })
      );
      break;

    case 'analyst_completed':
      // Update the specific analyst
      const analystIndex = newState.analysis.analysts.findIndex(
        (a) => a.chunkId === event.data.chunk_id
      );
      
      if (analystIndex !== -1) {
        newState.analysis.analysts = [...newState.analysis.analysts];
        newState.analysis.analysts[analystIndex] = {
          ...newState.analysis.analysts[analystIndex],
          status: 'complete',
          relevanceScore: event.data.relevance_score,
          isRelevant: event.data.is_relevant,
          summary: event.data.summary,
          confidence: event.data.confidence,
        };
      }
      
      newState.analysis.completedAnalysts = event.data.completed_analysts || 0;
      break;

    case 'synthesis_started':
      newState.status = 'synthesizing';
      newState.currentStep = 4;
      newState.synthesis.status = 'active';
      break;

    case 'final_response':
      newState.status = 'complete';
      newState.synthesis.status = 'complete';
      newState.totalDurationMs = event.data.total_duration_ms;
      newState.result = {
        finalResponse: event.data.final_response || '',
        retrievalCount: event.data.retrieval_count || 0,
        relevantCount: event.data.relevant_count || 0,
        sourcesUsed: event.data.sources_used || [],
        sharedDocument: event.data.shared_document || '',
      };
      break;

    case 'error':
      newState.status = 'error';
      newState.error = event.data.error;
      break;
  }

  return newState;
}

// Parse SSE data line
function parseSSEData(line: string): AgentEvent | null {
  if (!line.startsWith('data: ')) {
    return null;
  }
  
  try {
    const jsonStr = line.slice(6); // Remove 'data: ' prefix
    return JSON.parse(jsonStr) as AgentEvent;
  } catch (e) {
    console.error('Failed to parse SSE event:', e);
    return null;
  }
}

export interface StreamOptions {
  onEvent?: (event: AgentEvent) => void;
  onStateChange?: (state: PipelineState) => void;
  onComplete?: (state: PipelineState) => void;
  onError?: (error: Error) => void;
}

/**
 * Start a streaming RAG query with real-time progress updates.
 * 
 * @param query - The query to process
 * @param filters - Optional metadata filters
 * @param options - Callbacks for handling events
 * @returns AbortController to cancel the stream
 */
export async function startStreamingQuery(
  query: string,
  filters?: Record<string, unknown>,
  options?: StreamOptions
): Promise<AbortController> {
  const abortController = new AbortController();
  let state = createInitialPipelineState();

  try {
    const response = await fetch(`${API_BASE}/rag/stream`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'text/event-stream',
      },
      body: JSON.stringify({ query, filters }),
      signal: abortController.signal,
    });

    if (!response.ok) {
      throw new Error(`Stream request failed: ${response.status}`);
    }

    const reader = response.body?.getReader();
    if (!reader) {
      throw new Error('Response body is not readable');
    }

    const decoder = new TextDecoder();
    let buffer = '';

    // Process the stream
    (async () => {
      try {
        while (true) {
          const { done, value } = await reader.read();
          
          if (done) {
            options?.onComplete?.(state);
            break;
          }

          buffer += decoder.decode(value, { stream: true });
          const lines = buffer.split('\n');
          buffer = lines.pop() || ''; // Keep incomplete line in buffer

          for (const line of lines) {
            const trimmedLine = line.trim();
            if (!trimmedLine) continue;

            const event = parseSSEData(trimmedLine);
            if (event) {
              options?.onEvent?.(event);
              state = processAgentEvent(state, event);
              options?.onStateChange?.(state);

              if (state.status === 'complete' || state.status === 'error') {
                options?.onComplete?.(state);
              }
            }
          }
        }
      } catch (error) {
        if ((error as Error).name === 'AbortError') {
          // Stream was intentionally aborted
          return;
        }
        options?.onError?.(error as Error);
      }
    })();
  } catch (error) {
    options?.onError?.(error as Error);
  }

  return abortController;
}
